### PR TITLE
Only Cards mode for the Revit Plugin

### PR DESCRIPTION
**Pull request by aleixsuau**
https://community.openproject.com/projects/bcfier/work_packages/33494/activity

The main goal of this pull request is to create an "Only Cards" view model for the plugins environment (Revit). The following main changes have been made:

- [x] Modify router to be aware of the ViewerBrigde service in order to decide whether it should navigate to a route with a viewer ('bim.partitioned.split') or not ('bim.partitioned.list').
- [x] Modify the header's menu to only show the allowed display modes (only Cards in the plugins environment).
- [x] Test that the Cards display mode is showed by default on plugins environment and that there is no viewer and no viewer options on the menu.

### CODE REVIEW

**Comment by openprojectci** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/72cc3f39f0f437f1bfa9275d7428b59db8b5b3ee/frontend/src/app/components/op-context-menu/op-context-menu.component.ts#L7">frontend/src/app/components/op-context-menu/op-context-menu.component.ts#L7</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: config/middleware.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+require 'request_headers'
```
Comment (wielinde) [inline config/middleware.rb]: This file is obsolete as we now use a different approach to emulate a Revit addin environment in specs.

File: lib/request_headers.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+class CustomHeadersHelper
```
Comment (wielinde) [inline lib/request_headers.rb]: This file is also obsolete

File: frontend/src/app/modules/bim/ifc_models/toolbar/view-toggle/bim-view-toggle-dropdown.directive.ts [hunk 3]
```diff
@@ -68,35 +70,38 @@ export class BimViewToggleDropdownDirective extends OpContextMenuTrigger {
...
-              case bimViewerViewIdentifier:
-                this.state.go('bim.partitioned.model');
-                break;
-              case bimSplitViewIdentifier:
-                this.state.go('bim.partitioned.split');
-                break;
-            }
+                          switch (key) {
+                          case bimListViewIdentifier:
+                              this.state.go('bim.partitioned.list', {cards: true});
```
Comment (wielinde) [inline frontend/src/app/modules/bim/ifc_models/toolbar/view-toggle/bim-view-toggle-dropdown.directive.ts]: @aleixsuau We usually add a space after an opening curly brace and before an closing one. So `{ cards: true }` instead of `{cards: true}`.

File: frontend/src/app/components/op-context-menu/op-context-menu.html [hunk 4]
```diff
@@ -5,8 +5,8 @@
       role="menu"
       [ngClass]="{'-empty': items.length === 0 }">
     <ng-container *ngFor="let item of items">
-      <li *ngIf="item.divider && !item.hidden" class="dropdown-divider"></li>
-      <li *ngIf="!item.divider && !item.hidden">
+      <li *ngIf="item.divider" class="dropdown-divider"></li>
```
Comment (wielinde) [inline frontend/src/app/components/op-context-menu/op-context-menu.html:8]: @aleixsuau What has this todo with your PR? Did you come across this with Oliver? So are you sure it is safe to delete?

File: frontend/src/app/modules/bim/ifc_models/toolbar/view-toggle/bim-view-toggle-dropdown.directive.ts (hunk 3 above)
Comment (wielinde) [reply frontend/src/app/modules/bim/ifc_models/toolbar/view-toggle/bim-view-toggle-dropdown.directive.ts]: Just saw that we do not apply this in TypeScript... so please forget my comment

File: frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.ts [hunk 5]
```diff
@@ -26,14 +26,37 @@ import {IfcModelsDataService} from "core-app/modules/bim/ifc_models/pages/viewer
...
+
+    this.uIRouterGlobals
+          .params$!
+          .pipe(
+            this.untilDestroyed(),
+            pluck('cards'),
+            distinctUntilChanged(),
+            filter(cards => cards != null),
+          )
+          .subscribe((cards:boolean) => {
```
Comment (wielinde) [inline frontend/src/app/modules/bim/ifc_models/bcf/list-container/bcf-list-container.component.ts]: Isn't that logic that belongs to a different PR? We do not introduce a table in the BCF module in this PR, right? Or is that simply a preparation?

File: modules/bim/spec/features/bim_revit_addin_navigation_spec.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  it 'shows "Cards" view by default' do
+    model_page.expect_view_toggle_at 'Cards'
+  end
+
+  it 'shows no viewer' do
+    model_page.model_viewer_visible false
+  end
+
+  it 'has no viewer options' do
```
Comment (wielinde) [inline modules/bim/spec/features/bim_revit_addin_navigation_spec.rb:62]: For completeness you could extend this spec to also check that there is no "viewer + cards" item in it.

File: spec/support/capybara_headers.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+require 'request_headers'
```
Comment (wielinde) [inline spec/support/capybara_headers.rb]: This file is obsolete

File: frontend/src/typings.d.ts [hunk 8]
```diff
@@ -1,5 +1,7 @@
 /// <reference path="../tests/typings/tests.d.ts" />
-
+// Explicitly add UiRouterRx typings (in order to use params$ type)
```
Comment (wielinde) [inline frontend/src/typings.d.ts]: Here, you add a comment about UiRouterRx is that ever used?

File: frontend/src/app/components/op-context-menu/op-context-menu.html (hunk 4 above)
Comment (wielinde) [inline frontend/src/app/components/op-context-menu/op-context-menu.html:8]: This is a component that gets used in many places I believe. Are you sure that removing `&& !item.hidden` does not have side effects elsewhere?

File: config/middleware.rb (hunk 1 above)
Comment (aleixsuau) [reply config/middleware.rb]: Ok, removed

File: frontend/src/app/components/op-context-menu/op-context-menu.html (hunk 4 above)
Comment (aleixsuau) [reply frontend/src/app/components/op-context-menu/op-context-menu.html:8]: I think so. What I do in the component is to filter the items to those that are not hidden:
`this.items = this.locals.items.filter(item => !item?.hidden);` what it's basically the same that is done in the view.
I did that to get the '-empty' ngClass applied properly. It was not applied when there was only one item, but hidden, so the menu was still being opened even without visible items.

**Comment by openprojectci** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/8ecebf11b9da62b931acdcd6161ce0fa144933c0/frontend/src/app/components/op-context-menu/op-context-menu.component.ts#L7">frontend/src/app/components/op-context-menu/op-context-menu.component.ts#L7</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: lib/request_headers.rb (hunk 2 above)
Comment (aleixsuau) [reply lib/request_headers.rb]: Ok, deleted

File: modules/bim/spec/features/bim_revit_addin_navigation_spec.rb (hunk 6 above)
Comment (aleixsuau) [reply modules/bim/spec/features/bim_revit_addin_navigation_spec.rb:62]: Since the text option of have_no_selector looks for elements that **contains** the text, it should match both "Viewer" and "Viewer + Cards".

File: spec/support/capybara_headers.rb (hunk 7 above)
Comment (aleixsuau) [reply spec/support/capybara_headers.rb]: Ok, deleted

File: config/middleware.rb (hunk 1 above)
Comment (wielinde) [reply config/middleware.rb]: Hey Aleix, to me it seems this file is still here

**Review by wielinde — CHANGES_REQUESTED** [review]
Hey @aleixsuau, this looks all pretty good. There is just still one file that accidentally made it into the PR. 

Another note: We have many console.log messages in the revit bridge service. I think it makes sense to leave them in for the time being. However, we usually try not to be so noisy. Let's ask the others.

**Review by wielinde — APPROVED** [review]
LGTM
